File: Pattern-DynamicProgramming/LongestCommonSubstring/maximumSumIncreasingSubsequence/recursiveMemoization.js
// Given a number sequence, find the increasing subsequence with the highest sum. Write a method that returns the highest sum.

// Example 1:

// Input: {4,1,2,6,10,1,12}
// Output: 32
// Explanation: The increaseing sequence is {4,6,10,12}.
// Please note the difference, as the LIS is {1,2,6,10,12} which has a sum of '31'.

// Example 2:

// Input: {-4,10,3,7,15}
// Output: 25
// Explanation: The increaseing sequences are {10, 15} and {3,7,15}.

const findMSIS = function (nums) {
  const dp = [];
  const recursive = function (nums, previousIndex, currentIndex, sum) {
    if (currentIndex >= nums.length) {
      return sum;
    }
    const key = `${currentIndex}-${previousIndex}-${sum}`;
    if (typeof dp[key] === 'undefined') {
      let c1 = sum;
      if (previousIndex === -1 || nums[currentIndex] > nums[previousIndex]) {
        c1 = recursive(
          nums,
          currentIndex,
          currentIndex + 1,
          sum + nums[currentIndex]
        );
      }

      let c2 = recursive(nums, previousIndex, currentIndex + 1, sum);
      dp[key] = Math.max(c1, c2);
    }
    return dp[key];
  };
  return recursive(nums, -1, 0, 0);
};

console.log(
  `Maximum Sum Increasing Subsequence is: ---> ${findMSIS([
    4, 1, 2, 6, 10, 1, 12,
  ])}`
);
console.log(
  `Maximum Sum Increasing Subsequence is: ---> ${findMSIS([-4, 10, 3, 7, 15])}`
);

File: Pattern-DynamicProgramming/LongestCommonSubstring/subsequencePatternMatching/recursiveMemoization.py
# Given a string and a pattern, write a method to count the number of times the pattern appears in the string as a subsequence.

# Example 1: Input: string: “baxmx”, pattern: “ax”
# Output: 2
# Explanation: {baxmx, baxmx}.

# Example 2:

# Input: string: “tomorrow”, pattern: “tor”
# Output: 4
# Explanation: Following are the four occurences: {tomorrow, tomorrow, tomorrow, tomorrow}.

def find_SPM_count(str, pat):
    dp = [[-1 for _ in range(len(pat))] for _ in range(len(str))]
    return recursive(str,pat,0,0,dp)

def recursive(str,pat,strIdx, patIdx,dp):

    if patIdx == len(pat):
        return 1
    
    if strIdx == len(str):
        return 0

    if dp[strIdx][patIdx] == -1:


        c1 =0 
        if str[strIdx] == pat[patIdx]:
            c1 = recursive(str,pat,strIdx+1,patIdx+1,dp)
        
        c2 = recursive(str,pat,strIdx+1,patIdx,dp)

        dp[strIdx][patIdx] =  c2 + c1
    return dp[strIdx][patIdx]

def main():
  print(find_SPM_count("baxmx", "ax"))
  print(find_SPM_count("tomorrow", "tor"))


main()